src/main/java/org/jline/utils/AttributedStyle.java

```diff
@@ -42,14 +36,22 @@
     static final int F_CROSSED_OUT  = 0x00000080;
     static final int F_FOREGROUND   = 0x00000100;
     static final int F_BACKGROUND   = 0x00000200;
+    static final int F_GENERIC_ESCAPE  = 0x00400;
 
-    static final int MASK           = 0x000003FF;
+    static final int MASK           = 0x000007FF;
 
     static final int FG_COLOR_EXP    = 16;
     static final int BG_COLOR_EXP    = 24;
     static final int FG_COLOR        = 0xFF << FG_COLOR_EXP;
     static final int BG_COLOR        = 0xFF << BG_COLOR_EXP;
 
+    public static final AttributedStyle DEFAULT = new AttributedStyle();
+    public static final AttributedStyle BOLD = DEFAULT.bold();
+    public static final AttributedStyle BOLD_OFF = DEFAULT.boldOff();
+    public static final AttributedStyle INVERSE = DEFAULT.inverse();
+    public static final AttributedStyle INVERSE_OFF = DEFAULT.inverseOff();
+    public static final AttributedStyle GENERIC_ESCAPE =
+        new AttributedStyle(F_GENERIC_ESCAPE, F_GENERIC_ESCAPE);
 
```

Comment:
I don't think it's either possible or necessary to track styles in general, nor is there a need to split lines, at least not in a prompt.  In the degenerate case where a prompt must be printed across multiple lines, just do so.  I think we can safely assume that any terminal that supports styling, also supports ansi-style line-wrapping. Or at least we can refuse to support styling in that case.

I think you'd like a mechanism for "extended styling" that we can have an escape sequence to enter the mode, and one to exit the mode, so that if we need to split a line or only redraw a part of it then we can add extra exit/re-enter sequences.  But that's not always appropriate: not all styling is character attributes: some of it is structural.  Consider html:
  `<span class="foo">ABCXYZ</span>` is not necessarily the same as
 `<span class="foo">ABC</span><span="foo">XYZ</span>`
I want to be able to specify that `"ABCXYZ"` is unit of some kind, which is different from `"ABC"` followed by `"XYZ"`, so I don't want JLine to add extra escapes or groupings.  Especially when there is no need for it.

I'm aiming for some of functionality of IPython, where an application writes to a "scroll" - and the scroll can be saved, re-formatted, re-played, pretty-printed, etc etc. I'm not sure of all of the answers - perhaps we should just switch to IPython.  However, I'd like to be able to gradually increase sophistication and features, from very simple console input/output with basic editing to console complex user interfaces and more complex encodings of those interactions.  Concretely, I'm trying to design a simple general primitive protocol between an application and a smart console that can be used for more general structural encodings, and that JLine can "pass through" without munging or getting confused.